tmt/utils.py

```diff
@@ -2901,6 +2901,38 @@ def rewrite_git_url(url: str, patterns: List[Tuple[str, str]]) -> str:
     return url
 
 
+# Environment variable prefixes
+INJECT_CREDENTIALS_URL_PREFIX = 'TMT_GIT_CREDENTIALS_URL_'
+INJECT_CREDENTIALS_VALUE_PREFIX = 'TMT_GIT_CREDENTIALS_VALUE_'
+
+
+def inject_auth_git_url(url: str) -> str:
+    """
+    Inject username or token to the git url
+
+    :param url: original git repo url
+    :returns: URL with injected authentification based on pattern from the environment
+        or unmodified URL
+    """
+    # Try all environement variables sorted by their name
+    for name, value in sorted(os.environ.items(), key=lambda x: x[0]):
+        # First one which matches url is taken into the account
+        if name.startswith(INJECT_CREDENTIALS_URL_PREFIX) and re.search(value, url):
+            unique_suffix = name[len(INJECT_CREDENTIALS_URL_PREFIX):]
+            # Get credentials value
+            try:
+                creds = os.environ[f'{INJECT_CREDENTIALS_VALUE_PREFIX}{unique_suffix}']
+            except KeyError:
+                raise GitUrlError(f'Missing variable with credentials for "{url}"')
+            # Return original url if credentials is an empty value
+            if not creds:
+                return url
+            # Finally inject credentials into the url and return it
+            return re.sub(r'([^/]+://)([^/]+)', r'\1' + f'{creds}@' + r'\2', url)
```

Comment:
Cannot the replacement string be written together? Is the concatenation needed?

```suggestion
            return re.sub(r'([^/]+://)([^/]+)', rf'\1{creds}@\2', url)
```

Reply:
fixed in e7ac97ca